File: overlap.scala
case class Range(x: (Int, Int), y: (Int, Int)) extends Ordered[Range] {
  private def between(x: Int, lower: Int, upper: Int) =
    x >= lower && x <= upper

  private def overlaps(first: (Int, Int), second: (Int, Int)): Boolean = 
    between(first._1, second._1, second._2) ||
    between(first._2, second._1, second._2) ||
    between(second._1, first._1, first._2) ||
    between(second._2, first._1, first._2)

  private def combine(first: (Int, Int), second: (Int, Int)): (Int, Int) = 
    (first._1 min second._1, first._2 max second._2)

  def overlaps(other: Range): Boolean =
    overlaps(x, other.x) && overlaps(y, other.y)

  def combine(other: Range): Range =
    Range(combine(x, other.x), combine(y, other.y))

  def compare(other: Range) =
    if (x._1 == other.x._1 && x._2 == other.x._2) 0
    else if (x._1 < other.x._1 || (x._1 == other.x._1 && x._2 < other.x._2)) -1
    else 1
}

import scala.annotation.tailrec

@tailrec
def combine(input: List[Range], result: List[Range] = List.empty[Range]): List[Range] = {
  if (input.size < 2)
    return input ++ result

  val (x :: y :: tail) = input
  if (x overlaps y) 
    combine((x combine y) :: tail, result)
  else
    combine(y :: tail, x :: result)
}

val ranges = List(
  Range((0,6),   (34,40)),
  Range((1,7),   (35,41)),
  Range((3,9),   (12,18)),
  Range((2,8),   (36,42)),
  Range((13,19), (22,28)),
  Range((18,29), (23,30)))

val sorted = ranges.sorted
val combined = combine(sorted)
combined foreach println

File: overlap.hs
import Data.List

type Range = (Int, Int)
data RangePair = RangePair Range Range deriving (Eq, Ord, Show)

overlaps :: Range -> Range -> Bool
overlaps (a,b) (c,d) =
  c < a && a < d ||
  c < b && b < d ||
  a < c && c < b ||
  a < d && d < b

pairsOverlap :: RangePair -> RangePair -> Bool
pairsOverlap (RangePair a b) (RangePair c d) = (overlaps a c) && (overlaps b d)

combineRanges :: Range -> Range -> Range
combineRanges (a,b) (c,d) = (min a c, max b d)

combinePairs :: RangePair -> RangePair -> RangePair
combinePairs (RangePair a b) (RangePair c d) = RangePair (combineRanges a c) (combineRanges b d)

ranges = [
  RangePair (0,6)   (34,40),
  RangePair (1,7)   (35,41),
  RangePair (3,9)   (12,18),
  RangePair (2,8)   (36,42),
  RangePair (13,19) (22,28),
  RangePair (18,29) (23,30)]

combine :: [RangePair] -> [RangePair]
combine = map (foldl1 combinePairs) . groupBy pairsOverlap . sort
